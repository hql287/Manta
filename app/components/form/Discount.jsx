// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';

// Custom Components
import { Section } from '../shared/Section';

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Styles
import styled from 'styled-components';
const DiscountWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;
const DiscountContent = styled.div`
  display: flex;
  flex-direction: column;
`;
const DiscountAmount = styled.div`
  flex: 1;
`;
const DiscountType = styled.div`
  flex: 1;
  margin-top: 10px;
`;

// Component
export class Discount extends Component {
  constructor(props) {
    super(props);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.updateDiscountState = this.updateDiscountState.bind(this);
  }

  componentWillMount() {
    const { discount } = this.props;
    this.setState({
      amount: discount.amount ? discount.amount : '',
      type: discount.type ? discount.type : 'percentage',
    });
  }

  // Handle Reset Form
  componentWillReceiveProps(nextProps) {
    const { discount } = nextProps;
    if (isEmpty(discount)) {
      this.setState({
        amount: '',
        type: 'percentage',
      });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.state !== nextState || this.props.discount !== nextProps.discount
    );
  }

  handleInputChange(event) {
    const name = event.target.name;
    const eValue = event.target.value;
    let value;
    if (name !== 'amount') {
      value = eValue;
    } else {
      value = eValue === '' ? '' : parseInt(eValue, 10);
    }
    this.setState({ [name]: value }, () => {
      this.updateDiscountState();
    });
  }

  updateDiscountState() {
    const { updateFieldData } = this.props;
    updateFieldData('discount', this.state);
  }

  render() {
    return (
      <Section>
        <DiscountWrapper>
          <label className="itemLabel">Discount</label>
          <DiscountContent>
            <DiscountAmount>
              <input
                name="amount"
                type="number"
                value={this.state.amount}
                onChange={this.handleInputChange}
                placeholder="Amount"
              />
            </DiscountAmount>
            <DiscountType>
              <div className="radio">
                <label>
                  <input
                    name="type"
                    type="radio"
                    onChange={this.handleInputChange}
                    checked={this.state.type === 'percentage'}
                    value="percentage"
                  />Percentage
                </label>
              </div>
              <div className="radio">
                <label>
                  <input
                    name="type"
                    type="radio"
                    onChange={this.handleInputChange}
                    checked={this.state.type === 'flat'}
                    value="flat"
                  />Flat Rate
                </label>
              </div>
            </DiscountType>
          </DiscountContent>
        </DiscountWrapper>
      </Section>
    );
  }
}

Discount.propTypes = {
  discount: PropTypes.object.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Exports
export default _withFadeInAnimation(Discount);
