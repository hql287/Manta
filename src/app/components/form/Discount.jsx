// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';

// Custom Components
import { Section } from '../shared/Section';

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Styles
import styled from 'styled-components';
const DiscountWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;
const DiscountContent = styled.div`
  display: flex;
  flex-direction: column;
`;
const DiscountAmount = styled.div`
  flex: 1;
`;
const DiscountType = styled.div`
  flex: 1;
  margin-top: 10px;
`;

// Component
export class Discount extends Component {
  constructor(props) {
    super(props);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.updateDiscountState = this.updateDiscountState.bind(this);
  }

  componentWillMount() {
    const { discount } = this.props;
    this.setState({
      amount: discount.amount ? discount.amount : '',
      type: discount.type ? discount.type : 'percentage',
    });
  }

  // Handle Reset Form
  componentWillReceiveProps(nextProps) {
    const { discount } = nextProps;
    if (isEmpty(discount)) {
      this.setState({
        amount: '',
        type: 'percentage',
      });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.state !== nextState || this.props.discount !== nextProps.discount
    );
  }

  handleInputChange(event) {
    const name = event.target.name;
    const eValue = event.target.value;
    let value;
    if (name !== 'amount') {
      value = eValue;
    } else {
      value = eValue === '' ? '' : parseInt(eValue, 10);
    }
    this.setState({ [name]: value }, () => {
      this.updateDiscountState();
    });
  }

  updateDiscountState() {
    const { updateFieldData } = this.props;
    updateFieldData('discount', this.state);
  }

  render() {
    const { t } = this.props;
    return (
      <Section>
        <DiscountWrapper>
          <label className="itemLabel">
            {t('form:fields:discount.name')}
          </label>
          <DiscountContent>
            <DiscountAmount>
              <input
                name="amount"
                type="number"
                value={this.state.amount}
                onChange={this.handleInputChange}
                placeholder={t('common:amount')}
              />
            </DiscountAmount>
            <DiscountType>
              <div className="radio">
                <label>
                  <input
                    name="type"
                    type="radio"
                    onChange={this.handleInputChange}
                    checked={this.state.type === 'percentage'}
                    value="percentage"
                  />
                  {t('form:fields:discount:percentage')}
                </label>
              </div>
              <div className="radio">
                <label>
                  <input
                    name="type"
                    type="radio"
                    onChange={this.handleInputChange}
                    checked={this.state.type === 'flat'}
                    value="flat"
                  />
                  {t('form:fields:discount:flat')}
                </label>
              </div>
            </DiscountType>
          </DiscountContent>
        </DiscountWrapper>
      </Section>
    );
  }
}

Discount.propTypes = {
  discount: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Exports
export default _withFadeInAnimation(Discount);
