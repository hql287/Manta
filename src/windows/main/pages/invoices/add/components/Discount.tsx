// Libraries
import type React from 'react'
import { useState, useEffect } from 'react'
import type ChangeEvent from 'react'
import { isEmpty } from 'lodash'

// Custom Components
import { Section } from '@uiSharedComponents/Section'

// Animation
import _withFadeInAnimation from '@uiSharedComponents/hoc/_withFadeInAnimation'

// Styles
import styled from 'styled-components'

const DiscountWrapper = styled.div`
  display: flex;
  flex-direction: column;
`
const DiscountContent = styled.div`
  display: flex;
  flex-direction: column;
`
const DiscountAmount = styled.div`
  flex: 1;
`
const DiscountType = styled.div`
  flex: 1;
  margin-top: 10px;
`

// Type Definitions
interface DiscountProps {
  discount: {
    amount?: number
    type?: 'percentage' | 'flat'
  }
  t: (key: string) => string
  updateFieldData: (field: string, value: any) => void
}

const Discount: React.FC<DiscountProps> = ({
  discount,
  t,
  updateFieldData,
}) => {
  const [amount, setAmount] = useState<string | number>('')
  const [type, setType] = useState<'percentage' | 'flat'>('percentage')

  useEffect(() => {
    // Initialize state with discount props
    if (discount) {
      setAmount(discount.amount || '')
      setType(discount.type || 'percentage')
    }
  }, [discount])

  useEffect(() => {
    // Update discount state whenever `amount` or `type` changes
    updateFieldData('discount', { amount, type })
  }, [amount, type, updateFieldData])

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target

    if (name === 'amount') {
      setAmount(value === '' ? '' : parseInt(value, 10))
    } else if (name === 'type') {
      setType(value as 'percentage' | 'flat')
    }
  }

  return (
    <Section>
      <DiscountWrapper>
        <label className="itemLabel">{t('form:fields:discount.name')}</label>
        <DiscountContent>
          <DiscountAmount>
            <input
              name="amount"
              type="number"
              value={amount}
              onChange={handleInputChange}
              placeholder={t('common:amount')}
            />
          </DiscountAmount>
          <DiscountType>
            <div className="radio">
              <label>
                <input
                  name="type"
                  type="radio"
                  value="percentage"
                  onChange={handleInputChange}
                  checked={type === 'percentage'}
                />
                {t('form:fields:discount:percentage')}
              </label>
            </div>
            <div className="radio">
              <label>
                <input
                  name="type"
                  type="radio"
                  value="flat"
                  onChange={handleInputChange}
                  checked={type === 'flat'}
                />
                {t('form:fields:discount:flat')}
              </label>
            </div>
          </DiscountType>
        </DiscountContent>
      </DiscountWrapper>
    </Section>
  )
}

// Exports
export default _withFadeInAnimation(Discount)
