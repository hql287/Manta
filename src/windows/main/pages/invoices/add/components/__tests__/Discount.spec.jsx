// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

// Component
import { Discount } from '../Discount';

// Mocks
const t = jest.fn();
const updateFieldData = jest.fn();
const discount = {
  type: 'flat',
  amount: 100,
};

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Discount t={t} discount={discount} updateFieldData={updateFieldData} />
    );
  });

  // TODO
  it('will not rerender if threre is no change in Props', () => {});

  it('receives correct props', () => {
    expect(wrapper.prop('discount')).toEqual(discount);
    expect(wrapper.prop('discount')).not.toEqual({});
    expect(wrapper.prop('updateFieldData')).toEqual(updateFieldData);
  });

  it('renders necessary element', () => {
    expect(wrapper.find('label')).toHaveLength(3);
    expect(wrapper.find('input')).toHaveLength(3);
  });

  describe('handle input change correctly', () => {
    it('convert amount input to number', () => {
      // Setup
      const changeData = {
        target: {
          name: 'amount',
          value: '100',
        },
      };
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'handleInputChange');
      // Execute
      instance.handleInputChange(changeData);
      // Expect
      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(changeData);
      expect(updateFieldData).toHaveBeenCalledWith('discount', {
        type: 'flat',
        amount: 100,
      });
    });

    it('keep type input as string', () => {
      // Setup
      const changeData = {
        target: {
          name: 'type',
          value: 'percentage',
        },
      };
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'handleInputChange');
      // Execute
      instance.handleInputChange(changeData);
      // Expect
      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(changeData);
      expect(updateFieldData).toHaveBeenCalledWith('discount', {
        type: 'percentage',
        amount: 100,
      });
    });
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Discount t={t} discount={discount} updateFieldData={updateFieldData} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
