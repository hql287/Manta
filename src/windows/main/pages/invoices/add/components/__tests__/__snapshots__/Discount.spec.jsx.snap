// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="Section__SectionStyle-dgSKSv cwfAAu"
>
  <div
    className="Discount__DiscountWrapper-kWJKvg LseXW"
  >
    <label
      className="itemLabel"
    />
    <div
      className="Discount__DiscountContent-jCPLgI kgtKuP"
    >
      <div
        className="Discount__DiscountAmount-biDXgd btUZyY"
      >
        <input
          name="amount"
          onChange={[Function]}
          placeholder={undefined}
          type="number"
          value={100}
        />
      </div>
      <div
        className="Discount__DiscountType-gcYNnb jodsaL"
      >
        <div
          className="radio"
        >
          <label>
            <input
              checked={false}
              name="type"
              onChange={[Function]}
              type="radio"
              value="percentage"
            />
          </label>
        </div>
        <div
          className="radio"
        >
          <label>
            <input
              checked={true}
              name="type"
              onChange={[Function]}
              type="radio"
              value="flat"
            />
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
